crates/sound/src/device.rs

```diff
@@ -0,0 +1,717 @@
+                            return Err(Error::UnexpectedDescriptorSize(
+                                size_of::<VirtioSoundPcmStatus>(),
+                                descriptor.len(),
+                            )
+                            .into());
+                        }
+                        let mut streams = self.streams.write().unwrap();
+                        for b in std::mem::take(&mut buffers) {
+                            streams[stream_id as usize].buffers.push_back(b);
+                        }
+                        state = TxState::Done;
+                    }
+                    TxState::Ready
+                        if descriptor.len() as usize != size_of::<VirtioSoundPcmXfer>() =>
+                    {
+                        return Err(Error::UnexpectedDescriptorSize(
+                            size_of::<VirtioSoundPcmXfer>(),
+                            descriptor.len(),
+                        )
+                        .into());
+                    }
+                    TxState::Ready => {
+                        let xfer = desc_chain
+                            .memory()
+                            .read_obj::<VirtioSoundPcmXfer>(descriptor.addr())
+                            .map_err(|_| Error::DescriptorReadFailed)?;
+                        let stream_id: u32 = xfer.stream_id.into();
+                        stream_ids.insert(stream_id);
+
+                        state = TxState::WaitingBufferForStreamId(stream_id);
+                    }
+                    TxState::WaitingBufferForStreamId(stream_id)
+                        if descriptor.len() as usize == size_of::<VirtioSoundPcmXfer>() =>
+                    {
+                        return Err(Error::UnexpectedDescriptorSize(
+                            u32::from(
+                                self.streams.read().unwrap()[stream_id as usize]
+                                    .params
+                                    .buffer_bytes,
+                            ) as usize,
+                            descriptor.len(),
+                        )
+                        .into());
+                    }
+                    TxState::WaitingBufferForStreamId(_stream_id) => {
+                        let mut buf = vec![0; descriptor.len() as usize];
+                        let bytes_read = desc_chain
+                            .memory()
+                            .read(&mut buf, descriptor.addr())
+                            .map_err(|_| Error::DescriptorReadFailed)?;
+                        buf.truncate(bytes_read);
+
+                        buffers.push(Buffer::new(buf, Arc::clone(&message)));
+                    }
+                }
+            }
+        }
+
+        if !stream_ids.is_empty() {
+            let b = audio_backend.write().unwrap();
+            for id in stream_ids {
+                b.write(id).unwrap();
```

Comment:
If I understand correctly, in the tx queue, it collects all the data descriptors from a request. Then, It pushes for each descriptor that contains the data, the buffer and a reference to the message. When the status desc is identified, all the buffers are pushed back in the streams buffers queue. Then, for each `stream_id` that has been processed, it calls `write(stream_id)`. If you only have to notify guests after a `period_bytes` is consumed, would it not be better to only have a reference to the message for the set of buffers?

Reply:
> would it not be better to only have a reference to the message for the set of buffers?

The completion of a TX request can be considered an elapsed period notification but the reverse statement "tx requests must be completed (i.e. notify) after period_bytes are consumed" is not true.

In any case, the point of the references is reference count garbage collection; when all the buffers are dropped, the reference count of the IO message is zero so it is dropped as well, notifying the guest automatically.